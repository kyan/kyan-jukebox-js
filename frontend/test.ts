#!/usr/bin/env bun
/**
 * Simplified Bun Test Runner with Jest Compatibility
 *
 * Uses Bun's Jest compatibility mode to run existing Jest-style tests
 * without requiring complex migration of test files.
 */

import { spawnSync } from 'child_process'
import { existsSync } from 'fs'

interface TestOptions {
  watch?: boolean
  coverage?: boolean
  verbose?: boolean
  silent?: boolean
  bail?: boolean
  watchAll?: boolean
}

class BunJestRunner {
  private options: TestOptions = {}

  constructor(args: string[] = []) {
    this.parseArgs(args)
  }

  private parseArgs(args: string[]): void {
    for (let i = 0; i < args.length; i++) {
      const arg = args[i]

      switch (arg) {
        case '--watch':
        case '-w':
          this.options.watch = true
          break
        case '--watchAll':
          this.options.watchAll = true
          break
        case '--watchAll=false':
          this.options.watchAll = false
          this.options.watch = false
          break
        case '--coverage':
          this.options.coverage = true
          break
        case '--verbose':
          this.options.verbose = true
          break
        case '--silent':
          this.options.silent = true
          break
        case '--bail':
          this.options.bail = true
          break
        case '--forceExit':
        case '--detectOpenHandles':
        case '--maxWorkers=1':
          // Legacy Jest options - ignore as Bun handles these automatically
          break
      }
    }

    // Default watch behavior
    if (
      !process.env.CI &&
      this.options.watchAll === undefined &&
      this.options.watch === undefined
    ) {
      this.options.watch = true
    }
  }

  private setupEnvironment(): void {
    // Set test environment variables
    process.env.NODE_ENV = 'test'
    process.env.BABEL_ENV = 'test'
    process.env.PUBLIC_URL = ''
  }

  private async runPreflightChecks(): Promise<void> {
    if (this.options.silent) return

    console.log('üß™ Running preflight checks...\n')

    // Type checking
    if (existsSync('./tsconfig.json')) {
      console.log('üìù Type checking...')
      try {
        const typeCheck = spawnSync('bun', ['run', 'type-check'], {
          stdio: 'inherit'
        })

        if (typeCheck.status !== 0) {
          console.error('‚ùå Type check failed')
          process.exit(1)
        } else {
          console.log('‚úÖ Type check passed')
        }
      } catch (error) {
        console.warn('‚ö†Ô∏è Type checking skipped:', error.message)
      }
    }

    // Linting
    console.log('üîç Linting...')
    try {
      const lint = spawnSync('bun', ['run', 'lint'], {
        stdio: 'inherit'
      })

      if (lint.status !== 0) {
        console.error('‚ùå Linting failed')
        process.exit(1)
      } else {
        console.log('‚úÖ Linting passed')
      }
    } catch (error) {
      console.warn('‚ö†Ô∏è Linting skipped:', error.message)
    }

    // Format check
    console.log('üíÖ Format checking...')
    try {
      const format = spawnSync('bun', ['run', 'format:check'], {
        stdio: 'inherit'
      })

      if (format.status !== 0) {
        console.error('‚ùå Format check failed')
        console.log('üí° Run: bun run format')
        process.exit(1)
      } else {
        console.log('‚úÖ Format check passed')
      }
    } catch (error) {
      console.warn('‚ö†Ô∏è Format checking skipped:', error.message)
    }

    console.log('')
  }

  private buildJestCommand(): { command: string; args: string[] } {
    const args: string[] = []

    // Test pattern - Jest will find spec files automatically
    args.push('--testMatch', '<rootDir>/src/**/*.{spec,test}.{js,jsx,ts,tsx}')

    // Watch mode
    if (this.options.watchAll) {
      args.push('--watchAll')
    } else if (this.options.watch) {
      args.push('--watch')
    } else {
      args.push('--watchAll=false')
    }

    // Coverage
    if (this.options.coverage) {
      args.push('--coverage')
    }

    // Verbose mode
    if (this.options.verbose) {
      args.push('--verbose')
    }

    // Silent mode
    if (this.options.silent) {
      args.push('--silent')
    }

    // Bail on first failure
    if (this.options.bail) {
      args.push('--bail')
    }

    // Force exit and handle detection for CI
    if (process.env.CI || !this.options.watch) {
      args.push('--forceExit', '--detectOpenHandles', '--maxWorkers=1')
    }

    return { command: 'bunx', args: ['jest', ...args] }
  }

  async run(): Promise<void> {
    if (!this.options.silent) {
      console.log('üèÉ‚Äç‚ôÇÔ∏è Bun Jest Test Runner\n')
    }

    // Setup test environment
    this.setupEnvironment()

    // Run preflight checks (validation)
    await this.runPreflightChecks()

    // Build and run Jest command through Bun
    const { command, args } = this.buildJestCommand()

    if (this.options.verbose && !this.options.silent) {
      console.log('üìã Command:', [command, ...args].join(' '))
    }

    if (!this.options.silent) {
      console.log('üß™ Running tests...\n')
    }

    try {
      const result = spawnSync(command, args, {
        stdio: 'inherit',
        cwd: process.cwd(),
        env: {
          ...process.env,
          // Ensure Bun uses Jest compatibility mode
          FORCE_COLOR: '1'
        }
      })

      if (result.status !== 0) {
        if (!this.options.silent) {
          console.error('\n‚ùå Tests failed')
        }
        process.exit(result.status || 1)
      } else {
        if (!this.options.silent) {
          console.log('\n‚úÖ All tests passed!')
        }
      }
    } catch (error) {
      console.error('‚ùå Test runner failed:', error.message)

      if (!this.options.silent) {
        console.log('\nüí° Troubleshooting:')
        console.log('   ‚Ä¢ Ensure Bun v1.0+ is installed: bun --version')
        console.log('   ‚Ä¢ Try: bun install (to ensure dependencies are correct)')
        console.log('   ‚Ä¢ Check Jest configuration in package.json')
      }

      process.exit(1)
    }
  }
}

// CLI execution
if (import.meta.main) {
  const args = process.argv.slice(2)

  if (args.includes('--help') || args.includes('-h')) {
    console.log(`
Bun Jest Test Runner

Usage:
  bun test.ts [options]

Options:
  --watch, -w          Watch files for changes
  --watchAll           Watch all files
  --watchAll=false     Disable watch mode (for CI)
  --coverage           Generate coverage report
  --verbose            Verbose output
  --silent             Suppress output except errors
  --bail               Stop on first test failure

Examples:
  bun test.ts                           # Run all tests (watch mode in dev)
  bun test.ts --watchAll=false         # CI mode - run once
  bun test.ts --coverage               # With coverage report
  bun test.ts --silent --watchAll=false # Silent CI mode
`)
    process.exit(0)
  }

  const runner = new BunJestRunner(args)
  runner.run().catch(error => {
    console.error('Test runner failed:', error)
    process.exit(1)
  })
}
