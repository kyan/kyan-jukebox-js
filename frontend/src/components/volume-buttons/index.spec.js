import { describe, it, expect, mock } from 'bun:test'
import React from 'react'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import { render, fireEvent } from '@testing-library/react'
import VolumeButtons from './index'

describe('VolumeButtons', () => {
  const mockStore = configureMockStore()

  it('renders as expected', () => {
    const store = mockStore({
      jukebox: { volume: 50 }
    })

    const mockOnVolumeChange = mock(() => {})

    const { container } = render(
      <Provider store={store}>
        <VolumeButtons onVolumeChange={mockOnVolumeChange} />
      </Provider>
    )

    expect(container.firstChild).toBeInTheDocument()

    // Check that volume buttons are present
    const volumeDownButton = container.querySelector('.jb-volume-down')
    const volumeUpButton = container.querySelector('.jb-volume-up')

    expect(volumeDownButton).toBeInTheDocument()
    expect(volumeUpButton).toBeInTheDocument()
  })

  it('renders with disabled state', () => {
    const store = mockStore({
      jukebox: { volume: 50 }
    })

    const mockOnVolumeChange = mock(() => {})

    const { container } = render(
      <Provider store={store}>
        <VolumeButtons disabled={true} onVolumeChange={mockOnVolumeChange} />
      </Provider>
    )

    expect(container.firstChild).toBeInTheDocument()

    // Check that buttons are disabled
    const volumeDownButton = container.querySelector('.jb-volume-down')
    const volumeUpButton = container.querySelector('.jb-volume-up')

    expect(volumeDownButton).toBeDisabled()
    expect(volumeUpButton).toBeDisabled()
  })

  it('displays current volume', () => {
    const store = mockStore({
      jukebox: { volume: 75 }
    })

    const mockOnVolumeChange = mock(() => {})

    const { container } = render(
      <Provider store={store}>
        <VolumeButtons onVolumeChange={mockOnVolumeChange} />
      </Provider>
    )

    // Check that the volume is displayed in data-text attribute
    const volumeDisplay = container.querySelector('[data-text="75"]')
    expect(volumeDisplay).toBeInTheDocument()
  })

  it('calls onVolumeChange when volume up button is clicked', () => {
    const store = mockStore({
      jukebox: { volume: 50 }
    })

    const mockOnVolumeChange = mock(() => {})

    const { container } = render(
      <Provider store={store}>
        <VolumeButtons onVolumeChange={mockOnVolumeChange} />
      </Provider>
    )

    const volumeUpButton = container.querySelector('.jb-volume-up')
    fireEvent.click(volumeUpButton)

    expect(mockOnVolumeChange).toHaveBeenCalledWith(55)
  })

  it('calls onVolumeChange when volume down button is clicked', () => {
    const store = mockStore({
      jukebox: { volume: 50 }
    })

    const mockOnVolumeChange = mock(() => {})

    const { container } = render(
      <Provider store={store}>
        <VolumeButtons onVolumeChange={mockOnVolumeChange} />
      </Provider>
    )

    const volumeDownButton = container.querySelector('.jb-volume-down')
    fireEvent.click(volumeDownButton)

    expect(mockOnVolumeChange).toHaveBeenCalledWith(45)
  })

  it('does not call onVolumeChange when disabled', () => {
    const store = mockStore({
      jukebox: { volume: 50 }
    })

    const mockOnVolumeChange = mock(() => {})

    const { container } = render(
      <Provider store={store}>
        <VolumeButtons disabled={true} onVolumeChange={mockOnVolumeChange} />
      </Provider>
    )

    const volumeUpButton = container.querySelector('.jb-volume-up')
    const volumeDownButton = container.querySelector('.jb-volume-down')

    fireEvent.click(volumeUpButton)
    fireEvent.click(volumeDownButton)

    expect(mockOnVolumeChange).not.toHaveBeenCalled()
  })
})
