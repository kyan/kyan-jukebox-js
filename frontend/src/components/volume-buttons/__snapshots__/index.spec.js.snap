// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VolumeButtons render render 1`] = `
<VolumeButtons
  disabled={false}
  onVolumeChange={[MockFunction]}
>
  <ButtonGroup
    floated="right"
  >
    <div
      className="ui right floated buttons"
    >
      <VolumeDownButton
        disabled={false}
        onChange={[MockFunction]}
        volume={0}
      >
        <Button
          as="button"
          className="jb-volume-down"
          disabled={false}
        >
          <Ref
            innerRef={
              Object {
                "current": <button
                  class="ui button jb-volume-down"
                >
                  <i
                    aria-hidden="true"
                    class="volume down icon"
                  />
                </button>,
              }
            }
          >
            <RefFindNode
              innerRef={
                Object {
                  "current": <button
                    class="ui button jb-volume-down"
                  >
                    <i
                      aria-hidden="true"
                      class="volume down icon"
                    />
                  </button>,
                }
              }
            >
              <button
                className="ui button jb-volume-down"
                onClick={[Function]}
              >
                <Icon
                  as="i"
                  name="volume down"
                >
                  <i
                    aria-hidden="true"
                    className="volume down icon"
                    onClick={[Function]}
                  />
                </Icon>
              </button>
            </RefFindNode>
          </Ref>
        </Button>
      </VolumeDownButton>
      <span
        className="labelWrapper"
      >
        <ButtonOr
          text={0}
        >
          <div
            className="or"
            data-text={0}
          />
        </ButtonOr>
      </span>
      <VolumeUpButton
        disabled={false}
        onChange={[MockFunction]}
        volume={0}
      >
        <Button
          as="button"
          className="jb-volume-up"
          disabled={false}
          onClick={[Function]}
        >
          <Ref
            innerRef={
              Object {
                "current": <button
                  class="ui button jb-volume-up"
                >
                  <i
                    aria-hidden="true"
                    class="volume up icon"
                  />
                </button>,
              }
            }
          >
            <RefFindNode
              innerRef={
                Object {
                  "current": <button
                    class="ui button jb-volume-up"
                  >
                    <i
                      aria-hidden="true"
                      class="volume up icon"
                    />
                  </button>,
                }
              }
            >
              <button
                className="ui button jb-volume-up"
                onClick={[Function]}
              >
                <Icon
                  as="i"
                  name="volume up"
                >
                  <i
                    aria-hidden="true"
                    className="volume up icon"
                    onClick={[Function]}
                  />
                </Icon>
              </button>
            </RefFindNode>
          </Ref>
        </Button>
      </VolumeUpButton>
    </div>
  </ButtonGroup>
</VolumeButtons>
`;
